### ISSUE TITLE
Do not compile Python 3.9 with -fno-semantic-interposition flag

### ISSUE THREAD

**Issue by jhunkeler** [ContinuumIO/anaconda-issues#12319]
### Actual Behavior

Compiling C extensions for Python 3.9 under RHEL/Centos 7 is broken because GCC 4.8.5 does not support the `-fno-semantic-interposition` flag. This flag is not present in `python<3.9`.

### Expected Behavior

C extensions build successfully because a flag introduced by GCC 5.x isn't present.

### Steps to Reproduce

```$ conda create -n py39 python=3.9
$ conda activate py39
$ python3-config --cflags
-I/home/example/miniconda3/envs/bogus/include/python3.9 -I/home/example/miniconda3/envs/bogus/include/python3.9  -Wno-unused-result -Wsign-compare -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O3 -ffunction-sections -pipe -isystem /home/example/miniconda3/envs/bogus/include -fdebug-prefix-map=/tmp/build/80754af9/python-split_1614801158421/work=/usr/local/src/conda/python-3.9.2 -fdebug-prefix-map=/home/example/miniconda3/envs/bogus=/usr/local/src/conda-prefix -fno-semantic-interposition -fuse-linker-plugin -ffat-lto-objects -flto-partition=none -flto -DNDEBUG  -O3 -Wall

$ echo 'int main(){}' > test.c
$ gcc -fno-semantic-interposition test.c
gcc: error: unrecognized command line option ‘-fno-semantic-interposition’

$ gcc -v
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/libexec/gcc/x86_64-redhat-linux/4.8.5/lto-wrapper
Target: x86_64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-linker-build-id --with-linker-hash-style=gnu --enable-languages=c,c++,objc,obj-c++,java,fortran,ada,go,lto --enable-plugin --enable-initfini-array --disable-libgcj --with-isl=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/isl-install --with-cloog=/builddir/build/BUILD/gcc-4.8.5-20150702/obj-x86_64-redhat-linux/cloog-install --enable-gnu-indirect-function --with-tune=generic --with-arch_32=x86-64 --build=x86_64-redhat-linux
Thread model: posix
gcc version 4.8.5 20150623 (Red Hat 4.8.5-44) (GCC) 
```

##### Anaconda or Miniconda version:

```python-3.9.2-hdb3f193_0.conda```

##### `conda info`

<details>

Irrelevant

</details>

##### `conda list --show-channel-urls`

<details>

Irrelevant. It's the `defaults` channel set.

</details>

**Comment by mingwandroid** [comment]
We do want python compiled with this flag, we don't want external modules built with old GCC compiled with this flag.

Do you agree with this assessment or am I missing some details? This flag gives us a good speed boost.

**Comment by jhunkeler** [comment]
A colleague of mine pointed out it is possible to build the interpreter, libraries, and builtins _with_ `-fno-semantic-interposition` by adding it to `CFLAGS_NODIST` instead of `CFLAGS`:

```
$ ./configure [..] CFLAGS_NODIST="-fno-semantic-interposition"
```

This way the end-user still receives the speed boost at the interpreter level, with the added benefit of being able to use their OS's compiler for generic `python setup.py` and `pip install` calls during local development. CentOS 7 is documented as a supported operating system so its compiler should continue to be supported at least at some level. Do you agree?

Do you feel like this could be an acceptable compromise between speed and compatibility? If official packages compiled by `conda-build` want this flag too then perhaps it can test for its existence and append to its own `CFLAGS` accordingly.

```python
>>> import sysconfig
>>> for k,v in sysconfig.get_config_vars().items():
...     if "-fno-semantic-interposition" in str(v):
...             print(k)
... 
CONFIGURE_CFLAGS_NODIST
CONFIG_ARGS
PY_BUILTIN_MODULE_CFLAGS
PY_CFLAGS_NODIST
PY_CORE_CFLAGS
PY_STDMODULE_CFLAGS

>>> sysconfig.get_config_var("PY_CFLAGS_NODIST")
'-fno-semantic-interposition -std=c99 -Wextra -Wno-unused-result -Wno-unused-parameter -Wno-missing-field-initializers -Werror=implicit-function-declaration -fvisibility=hidden'```

**Comment by mingwandroid** [comment]
Yes. Thank you. I Will get right onto it! You're help is very much appreciated.

**Comment by jdavies-st** [comment]
We are seeing this same problem under Red Hat Enterprise Linux Server release 7.9 (Maipo) which has gcc 4.8.5, trying to build C extensions with the Anaconda distribution `python-3.9.4-hdb3f193_0`.  Same error.

**Comment by katietz** [comment]
I am taking right now a look to address this issue.  Thanks again @jhunkeler for your input.

**Comment by katietz** [comment]
Testing right now patch at https://github.com/AnacondaRecipes/python-feedstock/pull/29

**Comment by jdavies-st** [comment]
Thanks! Will updating the recipe (once merged) generate new builds of Python 3.9.2 and Python 3.9.4 packages?

**Comment by jhunkeler** [comment]
Did the testing not go well?

**Comment by katietz** [comment]
Yes, it works fine. Sorry, but we try to get out AIE release, so I post-spawned this patch for after it.  I will get to it for sure next week.

**Comment by jhunkeler** [comment]
@nden @jdavies-st Feel free to test the latest build of Python 3.9.5. I think it contains the proposed changes.

```
$ firewatch -t1c | grep ':\ python-3.9'
2020-10-22T08:35:32 : main/linux-64 : python-3.9.0-hcff3b4d_1.tar.bz2         
2020-11-15T09:41:14 : main/linux-64 : python-3.9.0-hdb3f193_2.tar.bz2         
2020-12-11T09:41:55 : main/linux-64 : python-3.9.1-hdb3f193_2.tar.bz2         
2021-03-03T15:14:38 : main/linux-64 : python-3.9.2-hdb3f193_0.tar.bz2         
2021-04-09T12:46:19 : main/linux-64 : python-3.9.4-hdb3f193_0.tar.bz2         
2021-05-18T15:46:41 : main/linux-64 : python-3.9.5-hdb3f193_3.tar.bz2 
```

**Comment by jdavies-st** [comment]
Yes, this is now fixed for the python 3.9.5 release.

🎉

**Comment by jhunkeler** [comment]
OK great! :partying_face: 
And thank you @katietz for making it happen. Much appreciated!
